// @tokenizer/s3/lib/s3-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const range_1 = require("@tokenizer/range");
/**
 * Use S3-client to execute actual HTTP-requests.
 */
class S3Request {
    constructor(s3, objRequest) {
        this.s3 = s3;
        this.objRequest = objRequest;
    }
    async getResponse(method, range) {
        const response = await this.getRangedRequest(range).promise();
        const contentRange = range_1.parseContentRange(response.ContentRange);
        return {
            size: contentRange.instanceLength,
            mimeType: response.ContentType,
            contentRange: contentRange,
            arrayBuffer: async () => {
                return response.Body;
            }
        };
    }
    /**
     * Do a ranged request
     * @param objRequest S3 object request
     * @param range Range request
     */
    getRangedRequest(range) {
        const rangedRequest = Object.assign({}, this.objRequest); // Copy request
        rangedRequest.Range = `bytes=${range[0]}-${range[1]}`;
        return this.s3.getObject(rangedRequest);
    }
}
exports.S3Request = S3Request;
